### ISSUE TITLE
Jump to the location where the snippet is defined

### ISSUE THREAD

**Issue by ZhiyuanLck** [SirVer/ultisnips#1198]
How can I jump to the location where the snippet is defined?
I know how to list the locations in the cmdline by
```python
:py3 print([(s.trigger, s.location) for s in UltiSnips_Manager._snips('', True)])
```
Is there a way to list them in the quickfix and when I press `Enter` I can jump to the corresponding line in the listed file or some other way like this?

**Comment by SirVer** [comment]
No, this is currently not possible, but would be a nice feature. I am not sure how the UI for this should look though: How would you select which snippet to jump to?

**Comment by ZhiyuanLck** [comment]
ui does not matter, either quickfix or popup is ok, I hope I can preview the snippet in popup or a split window and then choose whether to jump. And preview window can be toggle through a key or a pair of keys such as `p` and `P`.
Thanks a lot for reply.

**Comment by SirVer** [comment]
But which snippet should be shown?

**Comment by ZhiyuanLck** [comment]
In my case, suppose I have defined several snippets starting with `i`, then I type
```py
:py3 print([(s.trigger, s.location) for s in UltiSnips_Manager._snips('i', True)])
```
I will get all the results returned by a lists in a line.
```shell
[('it', '/home/zhiyuan/.vim/vim-init/UltiSnips/tex.snippets:93'), 
('int', '/home/zhiyuan/.vim/vim-init/UltiSnips/tex.snippets:240'), 
('item', '/home/zhiyuan/.vim/vim-init/UltiSnips/tex_refer.snippets:60'), 
('iff', '/home/zhiyuan/.vim/vim-init/UltiSnips/tex_refer.snippets:85'), 
('invs', '/home/zhiyuan/.vim/vim-init/UltiSnips/tex_refer.snippets:465'), 
('inn', '/home/zhiyuan/.vim/vim-init/UltiSnips/tex_refer.snippets:507')]
```
Maybe they could be displayed in some way like this in the quikfix
```shell
1. it      /home/zhiyuan/.vim/vim-init/UltiSnips/tex.snippets:93
2. int     /home/zhiyuan/.vim/vim-init/UltiSnips/tex.snippets:240
3. item    /home/zhiyuan/.vim/vim-init/UltiSnips/tex_refer.snippets:60
4. iff     /home/zhiyuan/.vim/vim-init/UltiSnips/tex_refer.snippets:85
5. invs    /home/zhiyuan/.vim/vim-init/UltiSnips/tex_refer.snippets:465
6. inn     /home/zhiyuan/.vim/vim-init/UltiSnips/tex_refer.snippets:507
```
I want to check/revise the definition of `int`, then I move to line `2` and press `<enter>` to open a new tab and jump to where the definition locates.

**Comment by ZhiyuanLck** [comment]
I realize that maybe there is no need to preview... but the way of preview could be the same as jump

**Comment by SirVer** [comment]
> :py3 print([(s.trigger, s.location) for s in UltiSnips_Manager._snips('i', True)])

Thanks, but that I have understood already. My questions are more about the intuitive user interface for this should look.

I still have trouble understanding the UI of this. What about regular expression triggers?

Basically we want to enter a trigger and have the snippets for that being available for editing, right? One way this could be done is by 

1. Entering the snippet into the buffer, then use a trigger that would not expand, but instead edit the snippet.
2. Query the user for a trigger interactively (not sure how?), then select and edit.

I see the appeal of the feature, but I have trouble wrapping my head around a potential implementation.

**Comment by ZhiyuanLck** [comment]
Yeah, the second way is what I said before but I am not good at vim script, so sorry for no help about the implementation. But if we use the first way you said, there may be some problems to be solved:
1. how do you recognize the snippet (such as an inner snippet) that does not exist?
2. For such a snippet `abc`, should it match one of the patterns `*abc`, `abc*`, `*abc*` or all, or even some fancy way (fuzzy match) `*a*b*c*`?
3. If it is not the exact match (just match `abc`) and there are many results, how do you choose which one to edit (maybe trigger a selection list to be chosen)?

**Comment by SirVer** [comment]
Shipped in #1668. Two entry points:

- `:UltiSnipsListLocations` populates the quickfix list with one entry per snippet definition and opens it via `:copen` - standard quickfix bindings (`<CR>`, `:cnext`, `:cprev`) jump to the matching file/line. Snippets registered through `UltiSnips#AddSnippetWithPriority` have no file location and are omitted.
- `UltiSnips#SnippetLocations()` returns the same data as a list of `{filename, lnum, text}` dicts (the `setqflist()` shape) for plugin authors who want to wire it into a fuzzy finder / popup / Telescope etc. instead of quickfix.

See `:h :UltiSnipsListLocations` and `:h UltiSnips#SnippetLocations`.
